### DISCUSSION TITLE
Mounting ASGI apps

### DISCUSSION THREAD

**Discussion by dantownsend** [Neoteroi/BlackSheep#160] [Ideas]
Is there a way to mount ASGI apps at a route? If not, this would be a nice feature.

```python
from blacksheep.server import Application

app = Application()

app.mount('/admin', some_other_asgi_app)
```

In FastAPI and [Starlette](https://www.starlette.io/routing/#submounting-routes) you can do this, which is a nice way of integrating third party apps into your own.

For example, I wrote this [admin GUI](https://github.com/piccolo-orm/piccolo_admin), which is intended to be mounted within other ASGI apps.

It's not a deal breaker, as there are work arounds, but would be nice to have if it's easy to add.

Also, congrats on the project - I came across it the other day, and the docs are great.

**Comment by RobertoPrevato** [comment]
Hi @dantownsend, thanks for your kind words!
`piccolo_admin` looks awesome. I wasn't considering support for mounting, yet, only because I'm still putting my effort on some "more basic" features that are missing from `blacksheep`, but I will put support for mounting among the items at the top of the TODO list (the backlog is handled in projects in GitHub).

**Comment by myusko** [comment]
Yeah, that's probably will be a good feature and promising, since it will allow us to cut off applications and potentially with different asgi frameworks into smaller parts, we can create an item in the backlog regarding this one.

**Reply by RobertoPrevato** [reply]
Sure, I just moved the existing work item into the private board for the web framework. Would you have time / would you like to look into this? I didn't check how it is implemented in Starlette, but if it's done as I suspect, it's using routing prefixes to delegate requests to the mounted app (in this case, it should be easy to add support for app mounting).

Now I would like to focus on updating the documentation, also documenting the `TestClient` (by the way, I didn't ask if you wish to write documentation for the `TestClient`, because I know usually this is not the most exciting things, while I have fun writing documentation).

**Reply by RobertoPrevato** [reply]
I just took a look here https://github.com/encode/starlette/blob/5ee04ef9b1bc11dc14d299e6c855c9a3f7d5ff16/starlette/routing.py#L307 - I confirm it's handled using prefixes and server side routing.

**Reply by myusko** [reply]
Yeah, that's would be nice if you can write documentation for the `TestClient` 😄 

Once, I've got a free time, I'll start work on the `mount` stuff. 
cc @dantownsend 

@RobertoPrevato BTW, I don't see a board with the stuff related to the web framework, only backlog.

**Comment by RobertoPrevato** [comment]
@dantownsend mount support is coming to BlackSheep, too! 🌞 

Here a screenshot of Piccolo Admin running under BlackSheep (still unpushed changes..):

![image](https://user-images.githubusercontent.com/2576032/123317516-671b8200-d52e-11eb-8d67-ee9757bd5ec5.png)

**Reply by dantownsend** [reply]
@RobertoPrevato Awesome, thanks for adding this. I think it's a really useful feature.

**Reply by RobertoPrevato** [reply]
And thanks to @myusko for his work! Thank You

**Reply by dantownsend** [reply]
Thanks @myusko !
